lighthouse-core/audits/byte-efficiency/unused-javascript.js

```diff
@@ -0,0 +1,133 @@
+/**
+ * @license
+ * Copyright 2017 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const ByteEfficiencyAudit = require('./byte-efficiency-audit');
+
+const IGNORE_THRESHOLD_IN_BYTES = 2048;
+
+class UnusedJavaScript extends ByteEfficiencyAudit {
+  /**
+   * @return {!AuditMeta}
+   */
+  static get meta() {
+    return {
+      category: 'Performance',
+      name: 'unused-javascript',
+      description: 'Unused JavaScript',
+      informative: true,
+      helpText: 'Remove unused JavaScript to reduce bytes consumed by network activity.',
+      requiredArtifacts: ['JsUsage', 'devtoolsLogs']
+    };
+  }
+
+  /**
+   * @param {{functions: !Array<{ranges: {startOffset: number, endOffset: number}}>}} script
+   * @return {{unusedLength: number, contentLength: number}}
+   */
+  static computeWaste(script) {
+    let maximumEndOffset = 0;
+    for (const func of script.functions) {
+      for (const range of func.ranges) {
+        maximumEndOffset = Math.max(maximumEndOffset, range.endOffset);
+      }
+    }
+
+    // We only care about unused ranges of the script, so we can ignore all the nesting and safely
+    // assume that if a range is unexecuted, all nested ranges within it will also be unexecuted.
+    const unusedByIndex = new Uint8Array(maximumEndOffset);
+    for (const func of script.functions) {
+      for (const range of func.ranges) {
+        if (range.count === 0) {
+          for (let i = range.startOffset; i < range.endOffset; i++) {
+            unusedByIndex[i] = 1;
```

Comment:
this seems really inefficient, but it's nice that it's so straightforward :) I guess we're only talking about 10s of MB even when uncompressed, so should be ok-ish until/if we hit some performance issues with it (or we want more out of coverage information)

Reply:
discussed in person, but it's actually much *more* efficient since flattening the many duplicated used ranges alone usually takes more time than iterating through the unused ranges (since there's only one range reporting the unused range)